### DISCUSSION TITLE
Mutate world from callbacks of third-party libraries

### DISCUSSION THREAD

**Discussion by mathisloge** [SanderMertens/flecs#1050] [Q&A] [answered]
Hi, 
I'm quite new to flecs and ecs in general so maybe I'm missing the point, but after reading the documentation I'm still unsure how to do certain things. 

In this case I'm using TGUI as my UI library and the scene example from https://github.com/SanderMertens/flecs/blob/master/examples/cpp/game_mechanics/scene_management/src/main.cpp
So when switching to my menu scene I'm creating the initial widgets and need to register the callbacks for the UI events
```cpp
gui->get<tgui::Button>("PlayButton")->onClick.connect([](auto) {
    // how to do this?
    world.add<ActiveScene, GameScene>();
});
```
In this case I want to switch to the game scene when the play button is pressed. 

Now I'm getting a little bit lost:
How can I access the world from inside the lamda? 
And when should I mutate the wold of the scene to not get some UB?

**Comment by SanderMertens** [comment]
You can capture variables in a lambda, so:
```cpp
flecs::world world;

// ...

gui->get<tgui::Button>("PlayButton")->onClick.connect([&](auto) { // note the [&]
    world.add<ActiveScene, GameScene>();
});
```

If you don't want modifications to take effect directly, you can put them between `defer_begin` and `defer_end`:

```cpp
world.defer_begin();
world.add<ActiveScene, GameScene>();
world.defer_end(); // add is executed here
```

**Reply by mathisloge** [reply]
In this case I can't capture by reference since the flecs::world will go out of scope.


```cpp
void menu_scene(flecs::iter &it, size_t, ActiveScene)
{
    std::cout << "\n>> ActiveScene has changed to `MenuScene`\n\n";
    flecs::world ecs = it.world();
    flecs::entity scene = ecs.component<SceneRoot>();

    reset_scene(ecs);
    ecs.get<Window>()->gui->loadWidgetsFromFile("/home/mathis/dev/project-world/game/main_menu.txt");
    ecs.get<Window>()->gui->get<tgui::Button>("PlayButton")->onClick.connect([&](const auto &) {
        ecs.add<ActiveScene, GameScene>();
    });

    ecs.set_pipeline(ecs.get<MenuScene>()->pip);
}
```
Sorry, I should have been posted the whole function at the beginning. 


As the copy constructor is deleted, I can't capture by value neither. Is there some way to get a not owned world instance into the lamda? Or is the way to go to create a class for the scences which are holding a reference to the world and the lamda just captures the class?

**Reply by SanderMertens** [reply, marked as answer]
Ah, that's a bit less straightforward.

What you can do is capture `flecs::world_t*` by value:

```cpp
flecs::world_t *ecs_ptr = ecs;

ecs.get<Window>()->gui->get<tgui::Button>("PlayButton")->onClick.connect([=](const auto &) {
  flecs::world ecs(ecs_ptr);
  // ...
});
```

I'll keep this use case in mind, I don't like how convoluted it is. Ideally the world object would be copyable, but that's a thorny problem to address (something to do with RAII and how copy constructors (don't) work).

**Reply by mathisloge** [reply]
Ah, thats working and definitly looks cleaner then the solutions I've come up with previously :D 
Thanks a lot for helping.
